betty/extension/webpack/build.py

```diff
@@ -0,0 +1,252 @@
+"""
+Perform Webpack builds.
+"""
+
+from __future__ import annotations
+
+from asyncio import to_thread
+from collections.abc import Sequence, MutableMapping
+from hashlib import md5
+from json import dumps, loads
+from logging import getLogger
+from os.path import join
+from pathlib import Path
+from shutil import copy2
+from typing import TYPE_CHECKING
+
+import aiofiles
+from aiofiles.os import makedirs
+
+from betty._npm import npm
+from betty.app.extension import Extension
+from betty.asyncio import gather
+from betty.cache.file import BinaryFileCache
+from betty.fs import ROOT_DIRECTORY_PATH, iterfiles
+from betty.job import Context
+from betty.locale import Localizer
+from betty.render import Renderer
+
+if TYPE_CHECKING:
+    from betty.extension.webpack import WebpackEntrypointProvider
+
+
+_NPM_INSTALL_DIRECTORY_PATH = Path(__file__).parent / "webpack"
+
+
+class BuildError(BaseException):
+    pass
+
+
+class NpmUnavailable(BuildError, RuntimeError):
+    pass
+
+
+class Builder:
+    def __init__(
+        self,
+        working_directory_path: Path,
+        entrypoint_providers: Sequence[WebpackEntrypointProvider & Extension],
+        debug: bool,
+        cache: BinaryFileCache,
+        renderer: Renderer,
+        *,
+        job_context: Context,
+        localizer: Localizer,
+    ) -> None:
+        self._working_directory_path = working_directory_path
+        self._entrypoint_providers = entrypoint_providers
+        self._debug = debug
+        self._cache = cache
+        self._renderer = renderer
+        self._job_context = job_context
+        self._localizer = localizer
+        self.__npm_install_id: str | None = None
+
+    async def _npm_install_key_for_package(self, npm_package_path: Path) -> str:
+        async with aiofiles.open(npm_package_path / "package.json") as f:
+            package_json = await f.read()
+        return md5(package_json.encode("utf-8")).hexdigest()
+
+    async def _npm_install_id(self) -> str:
+        if self.__npm_install_id is None:
+            self.__npm_install_id = md5(
```

Comment:
Can we turn this into a content hashing API? May be good for a separate PR as we have other code as well. Should that replace some of the existing base64 encoding we use?

Reply:
Should it replace fs.hashfile() too?

Reply:
Would be nice if the API could received structured keys, e.g. nested lists, and hash that too